pkg/xds/client/client.go

```diff
@@ -249,8 +256,76 @@ func (c *Client) handleResponse(log *slog.Logger, resp *discoverypb.DiscoveryRes
 			errs = errors.Join(errs, fmt.Errorf("missing name for typeUrl=%q", typeUrl))
 			continue
 		}
-		log.Debug("cache upsert", "name", name, "msg", msg)
-		c.cache.Upsert(typeUrl, name, msg)
+		if upsertedResources[typeUrl] == nil {
+			upsertedResources[typeUrl] = make(map[string]proto.Message)
+		}
+		upsertedResources[typeUrl][name] = msg
+	}
+	return upsertedResources, errs
+}
+
+func (c *Client) upsertAndDeleteMissing(ctx context.Context, upsertedResources map[string]map[string]proto.Message) error {
+	var errs error
+	for _, typeUrl := range []string{envoy.ListenerTypeURL, envoy.ClusterTypeURL, envoy.EndpointTypeURL} {
```

Comment:
Maybe the set of TypeURLs should be a property of the Client instead of being hardcoded here?